### PR TITLE
QAS Tickets

### PR DESCRIPTION
**Pull request by edwardstamps**
QAS Tickets with a focus around duplicate sessions, blank session, blank question or delete handlers

### CODE REVIEW

File: web/admin/src/Admin.js [hunk 1]
```diff
@@ -109,13 +115,33 @@ export default class Admin extends PureComponent {
...
   questionsForCurrentSession = () => Object.values(this.state.questionsBySession[this.state.sessionId] || {}).sort((a,b) => a.order - b.order)
 
   onSessionChange = e => this.setState({sessionId: e.target.value})
-  onSessionNameChange = e => this.sessionsRef().child(this.state.sessionId).update({name: e.target.value})
+
+  onSessionNameChange = e => {
+    const currentTitle = e.target.value.trim()
+    const dup = Object.values(this.state.sessions).find(i => i.name.toLowerCase() === currentTitle.toLowerCase())
+    if (currentTitle.length && !dup) {
+      this.publicSessionRef().child(this.state.sessionId).update({name: e.target.value, isDisplayable: true})
```
Comment (AdamLiechty) [inline web/admin/src/Admin.js]: It's better not to repeat these entire two lines for both cases.  The only thing different between lines 130/131 and 134/135 is the value of `isDisplayable`, so I would just have one copy of these two lines after assigning `isDisplayable` to `currentTitle.length > 0 && !dup`

File: web/admin/src/Admin.js [hunk 2]
```diff
@@ -109,13 +115,33 @@ export default class Admin extends PureComponent {
...
+    }
+    else {
+      this.publicSessionRef().child(this.state.sessionId).update({name: e.target.value, isDisplayable: false})
+      this.sessionsRef().child(this.state.sessionId).update({name: e.target.value, isDisplayable: false})
+    }
+  }
   onBackgroundUrlChange = e => this.backgroundUrlRef().set(e.target.value)
-  onSecondsChange = e => this.sessionsRef().child(this.state.sessionId).update({secondsPerQuestion: +e.target.value})
-  createSession = () => this.sessionsRef().push({name: 'New Session', secondsPerQuestion: 30}).then(ref => {
+  onSecondsChange = e => {
```
Comment (AdamLiechty) [inline web/admin/src/Admin.js]: IMO this should be a suggestion rather than forcing the event organizer to use max 60 seconds.  Someone will have a use case for longer.  Also, see https://stackoverflow.com/questions/8354975/how-can-i-limit-possible-inputs-in-a-html5-number-element#8355050

File: web/admin/src/Admin.js [hunk 3]
```diff
@@ -130,7 +156,17 @@ export default class Admin extends PureComponent {
   }
 
   addQuestion = (sessionId) => {
-    this.questionsRef().push({sessionId, order: this.questionsForCurrentSession().length, text: '', options: ['','','',''], correctIndex: 0})
+    const questions = this.questionsForCurrentSession()
+    if (questions.length) {
+      const checkQ = questions[questions.length-1]
+      if (checkQ.text && checkQ.options[0]) this.questionsRef().push({sessionId, order: this.questionsForCurrentSession().length, text: '', options: ['','','',''], correctIndex: 0})
```
Comment (AdamLiechty) [inline web/admin/src/Admin.js]: An `if/else` should at least indent the 2 branches, and preferably have braces around them for readability and preventing future logic errors.  It only adds 2 lines for the trailing braces.

File: web/admin/src/Admin.js [hunk 4]
```diff
@@ -27,7 +27,9 @@ export default class Admin extends PureComponent {
     sessions: {},
     questionsBySession: {},
     users: {},
-    publicSessions: {}
+    publicSessions: {},
+    helpText: "Add New Question",
```
Comment (AdamLiechty) [inline web/admin/src/Admin.js]: This `helpText` state variable seems like it has lots of places to end up in the wrong state, because the value it should have can be derived from other state variables, namely the existing questions and whether or not they are complete.  Any time you are trying to keep two pieces of state in sync, it is worth asking whether the render method can simply use one piece of state and derive the other.

File: web/admin/src/Questions.js [hunk 5]
```diff
@@ -38,7 +38,7 @@ export default class Questions extends PureComponent {
                       <div className="question-buttons">
                         { updater.state.hasPendingChanges && <button onClick={() => this.saveQuestion(item, updater)}>Save Question</button> }
                         { updater.state.hasPendingChanges && <button className="secondary" onClick={() => updater.cancel()}>Cancel Changes</button> }
-                        <button className="tertiary" onClick={this.deleteQuestion(item)}>Delete Question</button>
+                        <button className="tertiary" disabled={this.props.currentIndex === index} onClick={this.deleteQuestion(item)}>Delete Question</button>
```
Comment (AdamLiechty) [inline web/admin/src/Questions.js]: Why this change?

File: web/admin/src/Admin.js [hunk 6]
```diff
@@ -69,34 +72,37 @@ export default class Admin extends PureComponent {
...
+                  <div>Presentation Screen <button className="overlay-button" onClick={this.launchPresentation} disabled={launchDisabled || !this.bigScreenUrl()}>Launch Presentation</button></div>
                 </div>
-                <div className="presentation-side">
-                  <PresentationDriver fbc={this.props.fbc} session={sessions[sessionId]} questions={this.questionsForCurrentSession()} users={users} />
-                  <div className="presentation-overlays">
-                    <div>Up Next</div>
-                  </div>
+              </div>
+              <div className="presentation-side">
+                <PresentationDriver fbc={this.props.fbc} session={sessions[sessionId]} saveCurrentIndex={this.saveCurrentIndex} questions={this.questionsForCurrentSession()} users={users} />
```
Comment (AdamLiechty) [inline web/admin/src/Admin.js]: It looks like the index of the current question from the presentation (actually running a trivia game) is being sent out to the list of questions that can be edited.  Is this to prevent the CMS user from deleting a question that is currently showing in a trivia game?  If so, this will only prevent one case (you could delete it from another window, etc), and there are other weird things that can happen when deleting other questions for a game already in progress.  IMO I would not try to block such actions and complicate things.

File: web/admin/src/Admin.js [hunk 7]
```diff
@@ -69,34 +72,37 @@ export default class Admin extends PureComponent {
         { sessionId && <div>
             <label className="row">
               <span>Session Name:&nbsp;</span>
-              <input type="text" value={sessions[sessionId].name} onChange={this.onSessionNameChange} />
+              <input type="text" value={sessions[sessionId].name} maxLength={50} onChange={this.onSessionNameChange} />
+              {sessions[sessionId].isDisplayable ? null : <p>Please Rename Session</p>}
```
Comment (AdamLiechty) [inline web/admin/src/Admin.js]: `isDisplayable` is another new variable that can be derived from others and that can get out of sync.  If the name is empty, we don't need an extra `isDisplayable` flag... we know something is wrong with it.  If it is a duplicate, how do we determine which session's name should be changed?  I think it is fine to alert the CMS user that a name is duplicated (and this is derivable just by looking at the names of all the sessions, without a separate flag) but I would not add this flag and get the client side code in the business of hiding such sessions.

File: web/admin/src/Admin.js (hunk 2 above)
Comment (edwardstamps) [reply web/admin/src/Admin.js]: I agree. This makes the app less flexible.

File: web/admin/src/Admin.js (hunk 4 above)
Comment (edwardstamps) [reply web/admin/src/Admin.js]: Would you say the main issue is how resethelptext is called?

File: web/admin/src/Questions.js (hunk 5 above)
Comment (edwardstamps) [reply web/admin/src/Questions.js]: This prevents you from deleting the active question in a game. This caused issues with the index and also the question does not auto remove.

File: web/admin/src/Admin.js (hunk 6 above)
Comment (edwardstamps) [reply web/admin/src/Admin.js]: Would you suggest we prevent users from making any deletes during an active game?

File: web/admin/src/Admin.js (hunk 7 above)
Comment (edwardstamps) [reply web/admin/src/Admin.js]: Yes the purpose of this was two fold to alert the admin that a session is a duplicate or blank and hide it from the user. Should we add a pending state to handle incomplete questions or sessions?

File: web/admin/src/Admin.js (hunk 4 above)
Comment (AdamLiechty) [reply web/admin/src/Admin.js]: I would remove `resetHelpText()` and the `helpText` state altogether.  When it comes time to render, you can check the current questions to determine whether you want to show any help text.

File: web/admin/src/Admin.js (hunk 6 above)
Comment (AdamLiechty) [reply web/admin/src/Admin.js]: No... someone might decide that an upcoming question should not be shown at the last minute and delete it before we get there.

Comment (AdamLiechty) [reply web/admin/src/Admin.js]: Maybe we can disallow deleting the current question or any previous questions?

File: web/admin/src/Admin.js (hunk 7 above)
Comment (AdamLiechty) [reply web/admin/src/Admin.js]: I would avoid having a flag to signal that a question or session is incomplete. We should be able to tell just by looking at its fields when we want to know.  Having a separate field that tells whether it's incomplete duplicates data, and those duplicates can get out of sync.

File: web/admin/src/Admin.js (hunk 6 above)
Comment (edwardstamps) [reply web/admin/src/Admin.js]: To enable the fix in the above comment (disabling the button) Is there a better way to get the index of the current question in the game?

File: mobile/src/home-view.js [hunk 8]
```diff
@@ -126,6 +125,14 @@ export default class HomeView extends PureComponent {
       </View>
     )
   }
+
+  returnIsDisplayable = (s, sessions) => {
+    var isDisplayable = false
+    const currentTitle = s.name
+    const dup = Object.values(sessions).find(i => i.name.toLowerCase() === currentTitle.toLowerCase())
+    if (currentTitle.length && !dup) { isDisplayable = true }
```
Comment (AdamLiechty) [inline mobile/src/home-view.js]: Instead of initializing `isDisplayable` to false, then possibly setting to true, just say `const isDisplayable = currentTitle.length > 0 && !dup`

File: mobile/src/home-view.js [hunk 9]
```diff
@@ -126,6 +125,14 @@ export default class HomeView extends PureComponent {
       </View>
     )
   }
+
+  returnIsDisplayable = (s, sessions) => {
```
Comment (AdamLiechty) [inline mobile/src/home-view.js]: It's not idiomatic to call a function `returnXyz`.  Here, `isDisplayable` is a fine name for the function.  Note that this does not conflict with the variable with the same name inside the function, both because you are not calling the function from within itself, and because the function is being set as a property of `this` component (so it would have to be prefixed with `this.` to call it, anyway).

File: web/admin/src/Admin.js [hunk 10]
```diff
@@ -159,15 +158,25 @@ export default class Admin extends PureComponent {
...
+  returnHelpText = () => {
+    const questions = this.questionsForCurrentSession()
+    if (questions.length) {
+      const checkQ = questions[questions.length-1]
+      if (checkQ.text && checkQ.options[0]) {
+        return "Add New Question"
+      }
+      else return "Complete Last Question"
   }
+  else return "Add New Question"
```
Comment (AdamLiechty) [inline web/admin/src/Admin.js]: Tabs seem off here

File: web/admin/src/Admin.js (hunk 6 above)
Comment (AdamLiechty) [reply web/admin/src/Admin.js]: I think this turns out to be a fine way to bubble the index of the current question up.  It won't do anything about other tabs being open, but it will make it harder for the presenter to 👣 🔫

File: web/admin/src/Admin.css [hunk 11]
```diff
@@ -68,13 +68,13 @@ html, body, input, button, select, form{
   display: flex;
   position: relative;
 }
-@media screen and (min-width: 1230px) {
+/* @media screen and (min-width: 1230px) {
```
Comment (AdamLiechty) [inline web/admin/src/Admin.css]: Why remove this?

File: web/admin/src/Admin.js [hunk 12]
```diff
@@ -129,8 +129,9 @@ export default class Admin extends PureComponent {
 
   onSessionNameChange = e => {
     const currentTitle = e.target.value.trim()
-    this.publicSessionRef().child(this.state.sessionId).update({name: e.target.value})
+    const isPublicSession = Object.keys(this.state.publicSessions).find(item => item === this.state.sessionId)
```
Comment (AdamLiechty) [inline web/admin/src/Admin.js]: This seems like a convoluted way of saying `!!this.state.publicSessions[this.state.sessionId]`

Comment (edwardstamps) [reply web/admin/src/Admin.js]: Great point

File: web/admin/src/Admin.css (hunk 11 above)
Comment (edwardstamps) [reply web/admin/src/Admin.css]: Will would prefer the presentation view is always at the bottom. So this removes the shift from row to column.

Comment (AdamLiechty) [reply web/admin/src/Admin.css]: 👍
